#Bus Route and Link Extraction
with help of selenium webdriver though and by method only and i also used pandas for dataframing and time lib for loading content properly

In [1]:
from selenium import webdriver
from selenium.webdriver.common.by import By
import pandas as pd
import time



def extract_route(url):
    """Extracts route information from the provided URL."""
    driver.get(url)   
    time.sleep(5)  # Wait for the page to load
    try:
        driver.execute_script("window.scrollBy(0, 1600);")  # Scroll down to load elements
        time.sleep(2)  # Allow time for elements to load after scroll
        web_container = driver.find_element(By.CLASS_NAME, 'D117_main')
        page_tabs = web_container.find_elements(By.CLASS_NAME, 'DC_117_pageTabs')
        for page in page_tabs:
            page.click()  # Click on each page tab
            time.sleep(5)  # Wait for the page content to load
            elements = web_container.find_elements(By.CLASS_NAME, 'route_details')
            for element in elements:
                try:
                    anchor = element.find_element(By.TAG_NAME, 'a')
                    route_link['routename'].append(anchor.text)  # Store route name
                    route_link['routelink'].append(anchor.get_attribute('href'))  # Store route link
                except Exception as e:
                    pass  # Handle any exceptions that occur within the inner try block
            time.sleep(1)  # Brief pause before moving to the next set of elements
    except Exception as e:
        pass  # Handle any exceptions that occur within the outer try block

# List of URLs to extract route information from
route_dict = [
    'https://www.redbus.in/online-booking/apsrtc',
    'https://www.redbus.in/online-booking/tsrtc',
    'https://www.redbus.in/online-booking/rsrtc',
    'https://www.redbus.in/online-booking/hrtc',
    'https://www.redbus.in/online-booking/uttar-pradesh-state-road-transport-corporation-upsrtc',
    'https://www.redbus.in/online-booking/pepsu',
    'https://www.redbus.in/online-booking/pepsu-punjab',
    'https://www.redbus.in/online-booking/ksrtc-kerala',
    'https://www.redbus.in/online-booking/sikkim-nationalised-transport-snt',
    'https://www.redbus.in/online-booking/meghalaya-transport-corporation-mtc'
]


# Initialize dictionaries for storing route and bus details
driver = webdriver.Chrome()

route_link = {'routename': [], 'routelink': []}

# Extract route information from each URL in the list
for url in route_dict:
    extract_route(url)

driver.quit()  # Close the browser

# Convert the route data to a DataFrame and save it as a CSV file
route_data = pd.DataFrame(route_link)
route_data.to_csv('route_data.csv', index=False)



Route Name and Route Link Store in the *route data*

In [4]:
route_data

routename  \
0                   Vijayawada to Hyderabad   
1                   Hyderabad to Vijayawada   
2                 Kakinada to Visakhapatnam   
3                 Visakhapatnam to Kakinada   
4    Chittoor (Andhra Pradesh) to Bangalore   
..                                      ...   
274                  Shillong to Hailakandi   
275             Shillong to Ramkrishnanagar   
276                  Hailakandi to Shillong   
277             Ramkrishnanagar to Shillong   
278                    Shillong to Phulbari   

                                             routelink  
0    https://www.redbus.in/bus-tickets/vijayawada-t...  
1    https://www.redbus.in/bus-tickets/hyderabad-to...  
2    https://www.redbus.in/bus-tickets/kakinada-to-...  
3    https://www.redbus.in/bus-tickets/visakhapatna...  
4    https://www.redbus.in/bus-tickets/chittoor-and...  
..                                                 ...  
274  https://www.redbus.in/bus-tickets/shillong-to-...  
275  https://www.redbus.in/bus-tickets/shillong-to-...  
276  https://www.redbus.in/bus-tickets/hailakandi-t...  
277  https://www.redbus.in/bus-tickets/ramkrishnana...  
278  https://www.redbus.in/bus-tickets/shillong-to-...  

[279 rows x 2 columns]

that 279 link passed to the bus details extraction function, and collect the bus details

Bus Detials Extraction

In [5]:

# Initialize dictionary for storing bus details
bus_dict = {
    'route_Name':[],'busname': [], 'bus_desc': [],'bustype': [], 'starttime': [], 'endtime': [], 'duration': [],
    'dploc': [], 'ratings': [], 'fare': [], 'seats': [] }

# Initialize the WebDriver
driver = webdriver.Chrome()

def scroll(driver):
    """Scrolls to the bottom of the page to load dynamic content."""
    last_height = driver.execute_script("return document.body.scrollHeight")
    while True:
        driver.execute_script("window.scrollTo(0, document.body.scrollHeight);")
        time.sleep(3)
        new_height = driver.execute_script("return document.body.scrollHeight")
        if new_height == last_height:
            break
        last_height = new_height

def extract_time(name,url):
    """Extracts bus timing and other details from the provided URL."""
    driver.get(url)
    time.sleep(4)  # Wait for the page to load
    try:
        buttons = driver.find_elements(By.CLASS_NAME, 'p-left-10')
    except:
        buttons = []
    if buttons and len(buttons) > 1:
        for button in reversed(buttons):
            button.click()
            time.sleep(2)  # Allow time for content to load after click
    scroll(driver)
    time.sleep(10)  # Wait for content to load after scrolling
    web_container = driver.find_elements(By.CLASS_NAME, 'bus-item-details')
    for element in web_container:
        bus_dict["route_Name"].append(name)
        bus_dict['busname'].append(element.find_element(By.XPATH, "//div[@class='travels lh-24 f-bold d-color']").text)
        bus_dict['bus_desc'].append(element.find_element(By.CLASS_NAME, 'bus-type').text)
        bus_dict['bustype'].append(element.find_element(By.XPATH, "//div[@class='bus-type f-12 m-top-16 l-color evBus']").text)
        bus_dict['starttime'].append(element.find_element(By.CLASS_NAME, 'dp-time').text)
        bus_dict['endtime'].append(element.find_element(By.CLASS_NAME, 'bp-time').text)
        bus_dict['duration'].append(element.find_element(By.CLASS_NAME, 'dur').text)
        bus_dict['dploc'].append(element.find_element(By.CLASS_NAME, 'dp-loc').text)
        try:
            bus_dict['ratings'].append(element.find_element(By.CLASS_NAME, 'rating').find_element(By.TAG_NAME, 'span').text)
        except:
            bus_dict['ratings'].append("*")  # Default rating if not found
        bus_dict['fare'].append(element.find_element(By.CLASS_NAME, 'fare').text)
        bus_dict['seats'].append(element.find_element(By.CLASS_NAME, 'seat-left').text)

# List of route links (assuming route_link is already defined and populated)
for i, link in enumerate(route_link['routelink']):
    route_name = route_link['routename'][i]
    print(f"Processing route '{route_name}' ({i + 1}/{len(route_link['routelink'])}): {link} Total Extraction:{len(bus_dict["busname"])}")
    extract_time(route_name, link)


# Close the WebDriver
driver.quit()

# Convert the route data to a DataFrame and save it as a CSV file
bus_data = pd.DataFrame(bus_dict)
bus_data.to_csv('bus_data.csv', index=False)



Processing route 'Vijayawada to Hyderabad' (1/279): https://www.redbus.in/bus-tickets/vijayawada-to-hyderabad Total Extraction:0
Processing route 'Hyderabad to Vijayawada' (2/279): https://www.redbus.in/bus-tickets/hyderabad-to-vijayawada Total Extraction:342
Processing route 'Kakinada to Visakhapatnam' (3/279): https://www.redbus.in/bus-tickets/kakinada-to-visakhapatnam Total Extraction:729
Processing route 'Visakhapatnam to Kakinada' (4/279): https://www.redbus.in/bus-tickets/visakhapatnam-to-kakinada Total Extraction:731
Processing route 'Chittoor (Andhra Pradesh) to Bangalore' (5/279): https://www.redbus.in/bus-tickets/chittoor-andhra-pradesh-to-bangalore Total Extraction:734
Processing route 'Kadapa to Bangalore' (6/279): https://www.redbus.in/bus-tickets/kadapa-to-bangalore Total Extraction:744
Processing route 'Anantapur (andhra pradesh) to Bangalore' (7/279): https://www.redbus.in/bus-tickets/ananthapur-to-bangalore Total Extraction:785
Processing route 'Tirupati to Bangalore' 

WebDriverException: Message: disconnected: not connected to DevTools
  (failed to check if window was closed: disconnected: not connected to DevTools)
  (Session info: chrome=127.0.6533.89)
Stacktrace:
	GetHandleVerifier [0x00007FF62D2B9632+30946]
	(No symbol) [0x00007FF62D26E3C9]
	(No symbol) [0x00007FF62D166FDA]
	(No symbol) [0x00007FF62D14EA4C]
	(No symbol) [0x00007FF62D14E910]
	(No symbol) [0x00007FF62D1698C1]
	(No symbol) [0x00007FF62D1FA799]
	(No symbol) [0x00007FF62D1DC813]
	(No symbol) [0x00007FF62D1AA6E5]
	(No symbol) [0x00007FF62D1AB021]
	GetHandleVerifier [0x00007FF62D3EF83D+1301229]
	GetHandleVerifier [0x00007FF62D3FBDB7+1351783]
	GetHandleVerifier [0x00007FF62D3F2A03+1313971]
	GetHandleVerifier [0x00007FF62D2EDD06+245686]
	(No symbol) [0x00007FF62D27758F]
	(No symbol) [0x00007FF62D273804]
	(No symbol) [0x00007FF62D273992]
	(No symbol) [0x00007FF62D26A3EF]
	BaseThreadInitThunk [0x00007FF94841257D+29]
	RtlUserThreadStart [0x00007FF94A3CAA58+40]


In [40]:
bus_dict
bus_data = pd.DataFrame(bus_dict)

# DATA Cleaning

In [41]:
bus_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3608 entries, 0 to 3607
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   route_Name  3608 non-null   object
 1   busname     3608 non-null   object
 2   bus_desc    3608 non-null   object
 3   bustype     3608 non-null   object
 4   starttime   3608 non-null   object
 5   endtime     3608 non-null   object
 6   duration    3608 non-null   object
 7   dploc       3608 non-null   object
 8   ratings     3608 non-null   object
 9   fare        3608 non-null   object
 10  seats       3608 non-null   object
dtypes: object(11)
memory usage: 310.2+ KB


# Data Preprocessing Steps

1. **Clean and Convert Data Types**
   - **Rating, Fare, and Seats**: Ensure these columns are cleaned and converted to integer and float types as appropriate.
   - **Start Time and End Time**: Convert these columns to datetime data types.

2. **Separate Bus Name and Bus ID**
   - The `bus name` and `bus id` columns are currently combined. Separate them into distinct columns.

3. **Split Route Name**
   - **Route Name**: Split this column into two separate columns: `route_from` and `route_to`.

4. **Extract and Add Seat Type and AC Type**
   - **Seat Type and AC Type**: Extract these details from the `bus type` column and create separate columns for them.

By following these preprocessing steps, you will ensure that the data is clean, well-organized, and ready for further analysis.



In [42]:
df = bus_data.copy() # the original data copy taken
df

route_Name        busname  \
0       Vijayawada to Hyderabad  APSRTC - 3858   
1       Vijayawada to Hyderabad  APSRTC - 3858   
2       Vijayawada to Hyderabad  APSRTC - 3858   
3       Vijayawada to Hyderabad  APSRTC - 3858   
4       Vijayawada to Hyderabad  APSRTC - 3858   
...                         ...            ...   
3603  Vijayawada to Rajahmundry          NueGo   
3604  Vijayawada to Rajahmundry          NueGo   
3605  Vijayawada to Rajahmundry          NueGo   
3606  Vijayawada to Rajahmundry          NueGo   
3607  Vijayawada to Rajahmundry          NueGo   

                                            bus_desc  \
0     DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)   
1         Super Luxury (Non AC Seater 2+2 Push Back)   
2         AMARAVATHI (VOLVO / SCANIA A.C Multi Axle)   
3         Super Luxury (Non AC Seater 2+2 Push Back)   
4         Super Luxury (Non AC Seater 2+2 Push Back)   
...                                              ...   
3603                        A/C Seater/Sleeper (2+1)   
3604                               A/C Sleeper (2+1)   
3605                               A/C Sleeper (2+1)   
3606                               A/C Sleeper (2+1)   
3607                               A/C Sleeper (2+1)   

                                             bustype starttime endtime  \
0     DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)     11:30   16:30   
1     DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)     11:30   18:05   
2     DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)     12:15   18:25   
3     DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)     12:30   20:35   
4     DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)     13:00   21:00   
...                                              ...       ...     ...   
3603                       Electric A/C Seater (2+2)     22:20   02:00   
3604                       Electric A/C Seater (2+2)     22:30   02:50   
3605                       Electric A/C Seater (2+2)     23:05   03:30   
3606                       Electric A/C Seater (2+2)     23:45   03:15   
3607                       Electric A/C Seater (2+2)     00:55   04:15   

     duration                 dploc ratings      fare               seats  
0     05h 00m         Rtc bus stand     3.6   INR 670  26 Seats available  
1     06h 35m            Auto Nagar     4.0   INR 469  34 Seats available  
2     06h 10m         Rtc bus stand     3.8   INR 720  40 Seats available  
3     08h 05m            Auto Nagar     2.5   INR 469  28 Seats available  
4     08h 00m            Auto Nagar     2.0   INR 469  32 Seats available  
...       ...                   ...     ...       ...                 ...  
3603  03h 40m  Ankamma Thali Temple       *  INR 3999  44 Seats available  
3604  04h 20m             Tadepalli     2.7   INR 790  30 Seats available  
3605  04h 25m             Tadepalli     2.7   INR 890  24 Seats available  
3606  03h 30m             Tadepalli     2.7  INR 1090  28 Seats available  
3607  03h 20m             Tadepalli     2.7  INR 1290  20 Seats available  

[3608 rows x 11 columns]

# Step 1 : **Clean and Convert Data Types**
 - Rating - int type
 - fare - remove inr and float type
 - seats - remove seats avialble and convert number
 - starttime and end time - should data and time format

In [43]:
from datetime import datetime as dt

df['fare'] = df['fare'].str.replace('INR ', '')
df["fare"] = df["fare"].apply(lambda i : float(i))


df['ratings'] = df['ratings'].str.replace('*','0')
df["ratings"] = df["ratings"].apply(lambda i : float(i))

df['starttime'] = pd.to_datetime(df['starttime'], format='%H:%M')
df['endtime'] = pd.to_datetime(df['endtime'], format='%H:%M')

df['starttime'] = df['starttime'].dt.time
df['endtime'] = df['endtime'].dt.time

df.head(3)

route_Name        busname  \
0  Vijayawada to Hyderabad  APSRTC - 3858   
1  Vijayawada to Hyderabad  APSRTC - 3858   
2  Vijayawada to Hyderabad  APSRTC - 3858   

                                         bus_desc  \
0  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)   
1      Super Luxury (Non AC Seater 2+2 Push Back)   
2      AMARAVATHI (VOLVO / SCANIA A.C Multi Axle)   

                                          bustype starttime   endtime  \
0  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  11:30:00  16:30:00   
1  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  11:30:00  18:05:00   
2  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  12:15:00  18:25:00   

  duration          dploc  ratings   fare               seats  
0  05h 00m  Rtc bus stand      3.6  670.0  26 Seats available  
1  06h 35m     Auto Nagar      4.0  469.0  34 Seats available  
2  06h 10m  Rtc bus stand      3.8  720.0  40 Seats available

In [44]:
df[['Bus_Name', 'Bus_Id']] = df['busname'].str.split(' - ', expand=True)
df['Bus_Id'] = df['Bus_Id'].fillna('Private Bus')
df.loc[df['Bus_Id'] != 'Private Bus', 'Bus_Id'] = 'Government Bus'

df.head(3)

route_Name        busname  \
0  Vijayawada to Hyderabad  APSRTC - 3858   
1  Vijayawada to Hyderabad  APSRTC - 3858   
2  Vijayawada to Hyderabad  APSRTC - 3858   

                                         bus_desc  \
0  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)   
1      Super Luxury (Non AC Seater 2+2 Push Back)   
2      AMARAVATHI (VOLVO / SCANIA A.C Multi Axle)   

                                          bustype starttime   endtime  \
0  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  11:30:00  16:30:00   
1  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  11:30:00  18:05:00   
2  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  12:15:00  18:25:00   

  duration          dploc  ratings   fare               seats Bus_Name  \
0  05h 00m  Rtc bus stand      3.6  670.0  26 Seats available   APSRTC   
1  06h 35m     Auto Nagar      4.0  469.0  34 Seats available   APSRTC   
2  06h 10m  Rtc bus stand      3.8  720.0  40 Seats available   APSRTC   

           Bus_Id  
0  Government Bus  
1  Government Bus  
2  Government Bus

In [45]:
df['seats'] = df['seats'].str.replace('Seats available','')
df['seats'] = df['seats'].str.replace('Seat available','')
df["seats"] = df["seats"].apply(lambda i : int(i))

df.head(3)

route_Name        busname  \
0  Vijayawada to Hyderabad  APSRTC - 3858   
1  Vijayawada to Hyderabad  APSRTC - 3858   
2  Vijayawada to Hyderabad  APSRTC - 3858   

                                         bus_desc  \
0  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)   
1      Super Luxury (Non AC Seater 2+2 Push Back)   
2      AMARAVATHI (VOLVO / SCANIA A.C Multi Axle)   

                                          bustype starttime   endtime  \
0  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  11:30:00  16:30:00   
1  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  11:30:00  18:05:00   
2  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  12:15:00  18:25:00   

  duration          dploc  ratings   fare  seats Bus_Name          Bus_Id  
0  05h 00m  Rtc bus stand      3.6  670.0     26   APSRTC  Government Bus  
1  06h 35m     Auto Nagar      4.0  469.0     34   APSRTC  Government Bus  
2  06h 10m  Rtc bus stand      3.8  720.0     40   APSRTC  Government Bus

In [46]:
import re

def extract_seater(bustype):
    bustype_lower = bustype.lower()  # Convert to lowercase for consistent matching

    # Determine Seater type
    if 'seater / sleeper' in bustype_lower or 'seater/sleeper' in bustype_lower or \
       'sleeper / seater' in bustype_lower or 'sleeper/seater' in bustype_lower:
        seater = 'Seater / Sleeper'
    elif 'semi sleeper / sleeper' in bustype_lower or 'semi sleeper/sleeper' in bustype_lower or \
         'sleeper / semi sleeper' in bustype_lower or 'sleeper/semi sleeper' in bustype_lower:
        seater = 'Semi Sleeper / Sleeper'
    elif 'seater / semi sleeper' in bustype_lower or 'seater/semi sleeper' in bustype_lower or \
         'semi sleeper / seater' in bustype_lower or 'semi sleeper/seater' in bustype_lower:
        seater = 'Seater / Semi Sleeper'
    elif 'push back' in bustype_lower:
        seater = 'Push Back'
    elif 'semi sleeper' in bustype_lower:
        seater = 'Semi Sleeper'
    elif 'sleeper' in bustype_lower:
        seater = 'Sleeper'
    elif 'seater' in bustype_lower:
        seater = 'Seater'
    else:
        seater = 'Unknown'
    
    return seater

def extract_ac_type(bustype):
    bustype_lower = bustype.lower()  # Convert to lowercase for consistent matching

    # Regular expressions for matching Non-AC and AC variations
    ac_patterns = [r'\b(non[-\s]?ac)\b', r'\b(non[-\s]?a/c)\b', r'\b(non[-\s]?a\.c)\b', r'\bnon[-\s]?a\.c\b']
    non_ac_patterns = [r'\b(ac)\b', r'\ba/c\b', r'\ba\.c\b']

    ac_type = 'Unknown'  # Default value if no match is found

    # Check for Non-AC types first
    for pattern in ac_patterns:
        if re.search(pattern, bustype_lower):
            ac_type = 'Non A/C'
            break

    # Check for AC types if no Non-AC match was found
    if ac_type == 'Unknown':
        for pattern in non_ac_patterns:
            if re.search(pattern, bustype_lower):
                ac_type = 'AC'
                break

    return ac_type

# Applying the functions to your DataFrame
df['Seat_Type'] = df['bustype'].apply(extract_seater)
df['Actype'] = df['bustype'].apply(extract_ac_type)

df.head(5)


route_Name        busname  \
0  Vijayawada to Hyderabad  APSRTC - 3858   
1  Vijayawada to Hyderabad  APSRTC - 3858   
2  Vijayawada to Hyderabad  APSRTC - 3858   
3  Vijayawada to Hyderabad  APSRTC - 3858   
4  Vijayawada to Hyderabad  APSRTC - 3858   

                                         bus_desc  \
0  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)   
1      Super Luxury (Non AC Seater 2+2 Push Back)   
2      AMARAVATHI (VOLVO / SCANIA A.C Multi Axle)   
3      Super Luxury (Non AC Seater 2+2 Push Back)   
4      Super Luxury (Non AC Seater 2+2 Push Back)   

                                          bustype starttime   endtime  \
0  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  11:30:00  16:30:00   
1  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  11:30:00  18:05:00   
2  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  12:15:00  18:25:00   
3  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  12:30:00  20:35:00   
4  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  13:00:00  21:00:00   

  duration          dploc  ratings   fare  seats Bus_Name          Bus_Id  \
0  05h 00m  Rtc bus stand      3.6  670.0     26   APSRTC  Government Bus   
1  06h 35m     Auto Nagar      4.0  469.0     34   APSRTC  Government Bus   
2  06h 10m  Rtc bus stand      3.8  720.0     40   APSRTC  Government Bus   
3  08h 05m     Auto Nagar      2.5  469.0     28   APSRTC  Government Bus   
4  08h 00m     Auto Nagar      2.0  469.0     32   APSRTC  Government Bus   

  Seat_Type Actype  
0   Unknown     AC  
1   Unknown     AC  
2   Unknown     AC  
3   Unknown     AC  
4   Unknown     AC

In [47]:
df[['h', 'm']] = df['duration'].str.split(" ", expand=True)

df['h'] = df['h'].str.replace('h', '')
df['m'] = df['m'].str.replace('m', '')
df['h'] = df['h'].str.replace('-', '')
df['m'] = df['m'].str.replace('-', '')



In [48]:
# Convert to integers
df['h'] = df['h'].apply(lambda i: int(i))
df['m'] = df['m'].apply(lambda i: int(i))

df['dur_in_min'] = (df['h']*60) + df['m']
df = df.drop(columns=['h', 'm'])

In [49]:
df.head(5)

route_Name        busname  \
0  Vijayawada to Hyderabad  APSRTC - 3858   
1  Vijayawada to Hyderabad  APSRTC - 3858   
2  Vijayawada to Hyderabad  APSRTC - 3858   
3  Vijayawada to Hyderabad  APSRTC - 3858   
4  Vijayawada to Hyderabad  APSRTC - 3858   

                                         bus_desc  \
0  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)   
1      Super Luxury (Non AC Seater 2+2 Push Back)   
2      AMARAVATHI (VOLVO / SCANIA A.C Multi Axle)   
3      Super Luxury (Non AC Seater 2+2 Push Back)   
4      Super Luxury (Non AC Seater 2+2 Push Back)   

                                          bustype starttime   endtime  \
0  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  11:30:00  16:30:00   
1  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  11:30:00  18:05:00   
2  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  12:15:00  18:25:00   
3  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  12:30:00  20:35:00   
4  DOLPHIN CRUISE (VOLVO / SCANIA A.C Multi Axle)  13:00:00  21:00:00   

  duration          dploc  ratings   fare  seats Bus_Name          Bus_Id  \
0  05h 00m  Rtc bus stand      3.6  670.0     26   APSRTC  Government Bus   
1  06h 35m     Auto Nagar      4.0  469.0     34   APSRTC  Government Bus   
2  06h 10m  Rtc bus stand      3.8  720.0     40   APSRTC  Government Bus   
3  08h 05m     Auto Nagar      2.5  469.0     28   APSRTC  Government Bus   
4  08h 00m     Auto Nagar      2.0  469.0     32   APSRTC  Government Bus   

  Seat_Type Actype  dur_in_min  
0   Unknown     AC         300  
1   Unknown     AC         395  
2   Unknown     AC         370  
3   Unknown     AC         485  
4   Unknown     AC         480

In [37]:
df[['h', 'm']] = df['duration'].str.split(" ", expand=True)

In [50]:
df[['Orgin', 'End']] = df['route_Name'].str.split(' to ', expand=True)

In [51]:
df.to_csv("output.csv", index=False)

In [22]:
import sqlalchemy as sa
import mysql.connector

engine = sa.create_engine("mysql+mysqlconnector://admin:redbus123@redbus.c5ic8ayqm05z.ap-south-1.rds.amazonaws.com:3306/redbus")

data.to_sql('new',engine)

engine.dispose()